lib/index.ts

```diff
@@ -0,0 +1,494 @@
+                  require: require,
+                });
+              }
+            }
+            return plugin.hooks[name].map(function (hook) {
+              return merge(
+                {
+                  plugin: plugin.name,
+                  require: plugin.require,
+                },
+                hook
+              ) as Hook;
+            });
+          })
+          .filter(function (hook) {
+            return hook !== undefined;
+          })
+          .flat() as Hook[]),
+        ...(parent
+          ? parent.get({
+              name: name,
+              sort: false,
+            })
+          : []),
+      ];
+      if (!sort) {
+        return mergedHooks;
+      }
+      // Topological sort
+      const index: Record<string, Hook> = {};
+      for (const hook of mergedHooks) {
+        if (hook && 'plugin' in hook && hook.plugin) index[hook.plugin] = hook;
+      }
+      const edges_after = mergedHooks
+        .map(function (hook) {
+          if (!('after' in hook && Array.isArray(hook.after))) return;
+          return hook.after
+            .map(function (after) {
+              // This check assume the plugin has the same hooks which is not always the case
+              if (!index[after] && 'plugin' in hook && hook.plugin) {
+                if (registry.registered(after)) {
+                  throw errors.PLUGINS_HOOK_AFTER_INVALID({
+                    name: name,
+                    plugin: hook.plugin,
+                    after: after,
+                  });
+                } else {
+                  return undefined;
+                }
+              }
+              return [index[after], hook];
+            })
+            .filter(function (hook) {
+              return hook !== undefined;
+            }) as [Hook, Hook][];
+        })
+        .filter(function (hook) {
+          return hook !== undefined;
+        }) as [Hook, Hook][][];
+      const edges_before = mergedHooks
+        .map(function (hook) {
+          if (!('before' in hook && Array.isArray(hook.before))) return;
+          return hook.before
+            .map(function (before) {
+              if (!index[before] && 'plugin' in hook && hook.plugin) {
+                if (registry.registered(before)) {
+                  throw errors.PLUGINS_HOOK_BEFORE_INVALID({
+                    name: name,
+                    plugin: hook.plugin,
+                    before: before,
+                  });
+                } else {
+                  return undefined;
+                }
+              }
+              return [hook, index[before]];
+            })
+            .filter(function (hook) {
+              return hook !== undefined;
+            }) as [Hook, Hook][];
+        })
+        .filter(function (hook) {
+          return hook !== undefined;
+        }) as [Hook, Hook][][];
+      const edges = [...edges_after, ...edges_before].flat(0);
+      return toposort.array(mergedHooks, edges).map((hook) => {
```

Comment:
@wdavidw maybe you could help with this part ? I can't get the `edges` type right.

Reply:
I totally suck in Typescript, I had a look at it but no idea.

Reply:
I understand that it can be intimidating if your not used to the syntax, but what I'm asking is actually quite "simple" and doesn't require to fully understand typescript. You can approach it as any pseudo type language (lke the one you wrote in the readme).

You can also pull the PR locally to get some hints from your editor and eslint.

In fact I'm not sure if there is an issue with the typings, if I understood your code correctly, or if there is a bug in the original code.

Could you at least check lines L338 and L342 and tell me if `as [Hook, Hook][];` and `as [Hook, Hook][][];` are supposed to be the correct return types of the map+filter sequences ?

TS is complaining of receiving in the `edges` variable a `[Hook, Hook][][]` type while it expects a `[Hook, Hook | undefined][]` one. What make me think of an implementation issue is the additionnal `[]` array dimension that we have here.

Reply:
I did pull the changes and I confirm that I saw the TS error both in the CLI and in VSCode complaining about the `[Hook, Hook][][]`  and `[Hook, Hook | undefined][]`. However, I didn't know how to read those 2 last hints. I can give it another try

Reply:
The `as` keyword in `as [Hook, Hook][]` for example, is a type cast, to indicate to TS what the `hook.before(...).map(...).filter(...)` return type is supposed to be. Here I use it after the `filter` because typescript is not able to infer that the resulting array doesn't contain any `undefined` values anymore. If you hover the `filter` keyword, you'll see what is the actual type TS is inferring (`(Hook[] | undefined)[]` in this case).

Reply:
On L367 we end up with a `[Hook, Hook][][]`, which is correct in regards to the current implementation (please confirm),  but  we need a `[Hook, Hook][]`, as  `toposort.array` expects. We should have one nesting level less.

Reply:
I found a few things, tests are not passing but I got it to compile

- error.ts, line 2: remove `private`, error code is public
- index.ts, line 368: depth `0` in `array_flatten` convert to depth `1` in `flat`:
  ```ts
  const edges = [...edges_after, ...edges_before].flat(1);
  ```
- index.ts, line 369: pass the `Hook` trait and return `hook` from map
  ```ts
  return toposort.array<Hook>(mergedHooks, edges).map((hook) => {
    if (hook) {
      if ('require' in hook) delete hook.require;
      if ('plugin' in hook) delete hook.plugin;
    }
    return hook
  });
  ```

Reply:
Thank you, I will have a look

Reply:
> * index.ts, line 369: pass the `Hook` trait and return `hook` from map
>   ```ts
>   return toposort.array<Hook>(mergedHooks, edges).map((hook) => {
>     if (hook) {
>       if ('require' in hook) delete hook.require;
>       if ('plugin' in hook) delete hook.plugin;
>     }
>     return hook
>   });
>   ```

This is not required, [typescript infers the `Hook` type from arguments](https://www.typescriptlang.org/docs/handbook/2/generics.html).